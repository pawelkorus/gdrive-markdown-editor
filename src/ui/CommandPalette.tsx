import React, { useEffect, useState } from 'react';
import { Modal } from 'react-bootstrap';

export type Item = {
    id: string,
    name: string
}

interface CommandPaletteProps {
    show: boolean;
    commands: Item[]
    onItemSelected?: (commandPalleteItem: Item) => void
}

const CommandPalette: React.FC<CommandPaletteProps> = ({show, commands, onItemSelected}) => {
    const [isVisible, setIsVisible] = useState(show);
    const [selected, setSelected] = useState(0);

    useEffect(() => {
        setIsVisible(show);
    }, [show]);

    function onKeydown(e: React.KeyboardEvent<Element>) {
        const key = e.key;
        if (key === "ArrowDown") {
            e.preventDefault();
            setSelected((s) => {
                console.log(s);
                return Math.min(s + 1, commands.length - 1)
            });
            return;
        }
        if (key === "ArrowUp") {
            e.preventDefault();
            setSelected((s) => {
                console.log(s)
                return Math.max(s - 1, 0)
            });
            return;
        }
        if (key === "Enter") {
            e.preventDefault();
            if (onItemSelected) {
                onItemSelected(commands[selected]);
            }
            return;
        }
    }

    return (
        <Modal show={isVisible}>
            <Modal.Body onKeyDown={onKeydown}>
                <div className="list-group">
                {commands.map((item, i) => (
                        <a  key={i.toString()} 
                            href="#" 
                            className={`list-group-item list-group-item-action ${i === selected ? 'active' : ''}`} 
                            aria-current={i === selected ? true : false} 
                            onMouseDown={() => onItemSelected && onItemSelected(item)}
                            onMouseMove={() => setSelected(i)}
                            >
                            {item.name}
                        </a>
                        ))}
                </div>
            </Modal.Body>
        </Modal>
    );
};

export default CommandPalette;
