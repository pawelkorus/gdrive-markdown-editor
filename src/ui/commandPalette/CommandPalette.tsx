import React, { PropsWithChildren, useCallback, useEffect, useState } from 'react'
import { Modal, ModalBody, ModalHeader } from 'react-bootstrap'
import { CommandPaletteItem } from './types'

type CommandPaletteProps = PropsWithChildren<{
  commands: CommandPaletteItem[]
  onItemSelected?: (commandPalleteItem: CommandPaletteItem) => void
}>

const CommandPalette: React.FC<CommandPaletteProps> = ({ commands, children, onItemSelected }) => {
  const [isVisible, setIsVisible] = useState(false)
  const [lastShiftTimestamp, setLastShiftTimestamp] = useState(0)
  const [selected, setSelected] = useState(0)
  const [filteredCommands, setFilteredCommands] = useState([])
  const filterInputRef = React.createRef<HTMLInputElement>()

  const onKeyDown = useCallback((e: KeyboardEvent) => {
    const key = e.key

    if (key === 'ArrowDown' && isVisible) {
      e.preventDefault()
      setSelected((s) => {
        return Math.min(s + 1, commands.length - 1)
      })
      return
    }
    if (key === 'ArrowUp' && isVisible) {
      e.preventDefault()
      setSelected((s) => {
        return Math.max(s - 1, 0)
      })
      return
    }
    if (key === 'Enter' && isVisible) {
      e.preventDefault()
      handleCommandSelected(filteredCommands[selected])
      return
    }
    if (key === 'Shift' && !isVisible) {
      if (Date.now() - lastShiftTimestamp < 300) {
        e.preventDefault()
        setLastShiftTimestamp(0)
        setIsVisible(true)
      }
      else {
        setLastShiftTimestamp(Date.now())
      }
      return
    }
    if (key === 'Escape' && isVisible) {
      e.preventDefault()
      setIsVisible(false)
    }
  }, [selected, isVisible, filteredCommands, lastShiftTimestamp])

  useEffect(() => {
    setFilteredCommands(commands)
  }, [commands, isVisible])

  useEffect(() => {
    if (filterInputRef.current) {
      filterInputRef.current.focus()
    }
  }, [filterInputRef])

  useEffect(() => {
    document.addEventListener('keydown', onKeyDown)

    return () => {
      document.removeEventListener('keydown', onKeyDown)
    }
  }, [onKeyDown])

  function handleCommandSelected(item: CommandPaletteItem) {
    setIsVisible(false)
    if (onItemSelected) {
      onItemSelected(item)
    }
  }

  function filterItemsByName(e: React.ChangeEvent<HTMLInputElement>) {
    setFilteredCommands(commands.filter(command => command.name.toLowerCase().includes(e.target.value.toLowerCase())))
    setSelected(0)
  }

  return (
    <>
      {children}

      <Modal show={isVisible} fullscreen="sm-down">
        <ModalHeader>
          <input
            ref={filterInputRef}
            type="text"
            className="form-control"
            placeholder="Search..."
            onChange={filterItemsByName}
          />
        </ModalHeader>
        <ModalBody>
          <ul className="list-group">
            {filteredCommands.map((item, i) => (
              <li
                key={i.toString()}
                className={`list-group-item ${i === selected ? 'active' : ''}`}
                aria-current={i === selected ? true : false}
                onMouseDown={() => handleCommandSelected(item)}
                onMouseMove={() => setSelected(i)}
              >
                {item.name}
              </li>
            ))}
          </ul>
        </ModalBody>
      </Modal>
    </>
  )
}

export default CommandPalette
