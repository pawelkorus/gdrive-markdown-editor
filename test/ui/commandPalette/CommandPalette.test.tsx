import React from 'react';
import renderer from 'react-test-renderer';
// import { render, fireEvent } from '@testing-library/react';
import CommandPalette from '../../../src/ui/commandPalette/CommandPalette';
import { test, expect } from '@jest/globals';

test('renders hidden', () => {
    const commands = [{
        id: 'some-command',
        name: 'Some Command',
        execute: () => {}
    }];
    
    const component = renderer.create(<CommandPalette commands={commands}/>).toJSON();

    expect(component).toMatchSnapshot();
});
