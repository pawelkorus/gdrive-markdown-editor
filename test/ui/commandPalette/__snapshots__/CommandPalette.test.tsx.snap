// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders hidden 1`] = `null`;
